spec/abci/abci++_basic_concepts.md

```diff
@@ -200,25 +204,27 @@ and an [**Echo**](./abci++_methods.md#echo) method that is used for debugging.
 More details on managing state across connections can be found in the section on
 [Managing Application State](./abci%2B%2B_app_requirements.md#managing-the-application-state-and-related-topics).
 
-## Proposal timeout
+## Proposal related timeouts
 
-Immediate execution requires the Application to fully execute the prepared block
-before returning from `PrepareProposal`, this means that CometBFT cannot make progress
-during the block execution.
-This stands on the consensus algorithm critical path: if the Application takes a long time
-executing the block, the default value of *TimeoutPropose* might not be sufficient
-to accommodate the long block execution time and non-proposer nodes might time
-out and prevote `nil`. The proposal, in this case, will probably be rejected and a new round will be necessary.
+`PrepareProposal` and `ProcessProposal` stand on the consensus algorithm critical path,
+i.e., CometBFT cannot make progress while these methods are being executed.
+Hence, if the Application takes a long time preparing a proposal or in verifying it,
+the default values of *TimeoutPropose* and *TimeoutPrevote* might not be sufficient
+to accommodate these methods' executions and validator nodes might time out and prevote `nil`.
+The proposal, in this case, will probably be rejected and a new round will be necessary.
+If this happens successively, the liveness of CometBFT might be compromised.
```

Comment:
If the implementation of `PrepProp` and `ProcProp` are bounded in time, we don't introduce liveness problems (the algorithm is guaranteed to terminate), but performance problems (unwanted rounds).

Reply:
> Ah, this is a very important information. What happens if the application doesn't reply on time? What is proposed?

CometBFT blocks until the application replies (similarly to `FinalizeBlock`). It's a **sync** ABCI call

Reply:
So, timeout prevote does not apply here, as it is only triggered when the node receives 2/3+ votes that do not form a quorum.

Reply:
> I still feel we shouldn't say that liveness is compromised (but performance is degraded). Thoughts?

I think it affects liveness somehow. The point is that timeouts are increasing, so eventually they become big enough so that to include the processing time involved on these calls.

Reply:
> > I still feel we shouldn't say that liveness is compromised (but performance is degraded). Thoughts?
> 
> I think it affects liveness somehow. The point is that timeouts are increasing, so eventually they become big enough so that to include the processing time involved on these calls.

Good point. At some point then proposals will be made before the other validators prevote nil and eventually we decide. I agree with you that it is not a liveness issue, but a performance one. A potentially very painful one, since the timeout values are reset for each new round.